README.md
# Terminal Clone

Welcome to Terminal Clone, a versatile command-line interface that replicates terminal functionality. Explore commands for system management, internet interaction, and Python package handling. Learn about available actions and their usage below.

**Please note that this project is a work in progress and is not yet a complete demo release. Some features might not be fully functional or included in this GitHub repository at the moment, but we have plans to add them soon.**

*A release date marks the day when specific features and bug fixes are made available. Not all updates may be present now, but rest assured, we're actively working to update them in the near future.*

## Table of Contents
- [Login Command](#login-command)
- [System Command](#system-command)
- [Internet Command](#internet-command)
- [PPM (Python Package Manager) Command](#ppm-python-package-manager-command)
- [Release Version](#Release-Version)
- [Recommended Python Version](#Recommended-Python-Version)
- [About Us](#about-us)



## Login Command

This set of commands allows you to manage user authentication and login state.

- `.register`: Create a new account.
- `.login`: Log in to your account.
- `.state`: Show the current login state.
- `.logout`: Log out from your account.

Your user data is securely stored locally at `source/users.txt` and **will not** be sent to any external servers.

## System Command

Use these commands to interact with system-related operations. Please note that all of the following actions are simulated and **do not affect your actual system**.

- `.help`: Get information about available commands.
- `.test`: Check system processes, files, or encryption.
- `.render`: Open and render a 3D model to assess performance.
- `.exit`: Safely exit the terminal program.

## Internet Command

Interact with online services using these commands.

- `.connect`: Connect to a server to access various services.
  
**Please keep in mind:**

- **Real Connection Check**: The `.connect` command checks your internet connection using the Python `socket` library, without sending actual data.

- **Educational Use**: These commands are for learning and demonstration purposes, not actual internet access or activities.

## PPM (Python Package Manager) Command

Manage Python packages using the PPM command. Please note that this command is for demonstration purposes and **does not perform actual installations**.

- `ppm`: Open the PPM function.

    The PPM (Python Package Manager) command provides a simple way to **simulatedly** manage Python packages within this terminal clone. You can use this command to install, update, or remove packages for your projects. Please keep in mind that the installations are not real and only serve as examples.

    After running the PPM command, you might notice the creation of a `__pycache__` directory. This directory stores compiled Python bytecode files and is automatically generated by Python for performance optimization. It is safe to delete the `__pycache__` directory if you no longer need it. Deleting it will not have any impact on the functionality of this terminal clone or your projects. The directory will be regenerated as needed.

- `ppm --list`: List installed packages on this PC.

    Use this command to view a list of Python packages currently installed on your PC.

Please feel free to manage your Python packages using the PPM command, knowing that the package installations are not actual and that deleting the `__pycache__` directory won't affect the functionality of this terminal clone or your projects.

## Release Version

**Current Release Version: 3.3**

**Release Date: August 20, 2023**

We're excited to introduce the initial release of Terminal Clone, version 3.3 This marks an important milestone in our project's development, bringing a range of features and functionalities to provide a comprehensive command-line experience. While this release is not the final product, it demonstrates our commitment to creating a versatile and user-friendly terminal simulation.

For detailed release notes and changes, please refer to the [changelog](https://github.com/donut9164/Terminal-clone/blob/main/changelog.txt) file.

### Future Plans and Feedback

We're dedicated to continuing the development of Terminal Clone and expanding its capabilities in upcoming releases. Your feedback is invaluable to us, and we encourage you to share your thoughts, suggestions, and ideas for improvement. Together, we can shape the future of this project and create an even more powerful and educational tool.

Stay tuned for future updates and releases as we work towards providing a comprehensive and interactive terminal experience.

Thank you for being a part of the Terminal Clone community!

## Recommended Python Version

We strongly recommend using Python version 3.10.xx for the best experience with this project. While efforts are made to ensure compatibility with other versions, Python 3.10.xx offers optimal performance and compatibility with the features of this project.

Python 3.10.xx introduces numerous enhancements, optimizations, and new features that can greatly benefit your usage of this project. To install Python 3.10.xx, you can visit the [official Python website](https://www.python.org/downloads/) and follow the installation instructions for your platform.

Please note that while this project does support Python version 3.11.x, we do not recommend its use due to potential performance issues. Some features might not be fully optimized or compatible with Python 3.11.x We are actively working to address these issues and plan to provide better support for Python 3.11.x and other versions in the future.

Staying up to date with the latest releases of Python is advised for the most secure and optimized development experience.

## About Us

Learn more about the creators of this terminal clone.

- `.about`: Information about our team and project.
- `.contact`: Find out how to get in touch with us.

---

## Feel Free to Edit and Configure

We encourage you to tailor this terminal clone to your preferences. Feel free to edit the commands, add new functionalities, and configure it according to your needs. To get started, you can explore the source code and make modifications as desired.

Feel the power of customization and make this terminal clone your own! If you have any questions or ideas, don't hesitate to [reach out to us](). 


changelog.txt
Changelog
----------

Version 3.3 - August 20, 2023
- Addressed 3D rendering performance issues.
- Ongoing work to resolve PPM installation errors (work in progress).

Version 3.2 - August 14, 2023
- Included additional packages to PPM (Python Package Manager).

Version 3.1 - August 14, 2023
- Addressed 3D rendering performance issues.
- Ongoing work to resolve PPM installation errors (work in progress).

Version 3.0 - August 14, 2023
- Introduced PPM (Python Package Manager) function.
- New command: `.render` for 3D model rendering.

Version 2.1 - August 9, 2023
- Fixed bug preventing internet connection.
- Addressed connectivity issues.
- Optimized performance.

Version 2.0 - August 5, 2023
- Added login command.
- Added internet commands: `.connect`.

Version 1.3 - July 31, 2023
- Resolved several bugs.
- Fixed issues with `.test` command execution.

Version 1.2 - July 29, 2023
- Addressed various minor bugs.

Version 1.1 - July 27, 2023
- Fixed specific bugs.

Version 1.0 - September 24, 2023
- Added basic commands.
- Introduced `.test` command for system testing.


3d.py
import tkinter as tk
import math
import time

print("Running at")

class SpinningCubeApp:
    def __init__(self, root):
        self.root = root
        self.root.title("3D Render")

        self.canvas = tk.Canvas(root, width=400, height=400)
        self.canvas.pack()

        self.cube_vertices = [
            (-50, -50, -50),
            (50, -50, -50),
            (50, 50, -50),
            (-50, 50, -50),
            (-50, -50, 50),
            (50, -50, 50),
            (50, 50, 50),
            (-50, 50, 50)
        ]

        self.cube_edges = [
            (0, 1), (1, 2), (2, 3), (3, 0),
            (4, 5), (5, 6), (6, 7), (7, 4),
            (0, 4), (1, 5), (2, 6), (3, 7)
        ]

        self.angle_x = 0
        self.angle_y = 0
        self.last_frame_time = time.time()  # Track time of last frame
        self.frames = 0  # Frame counter
        self.animate()

    def rotate_point(self, point, angle_x, angle_y):
        x, y, z = point

        cos_x = math.cos(angle_x)
        sin_x = math.sin(angle_x)
        cos_y = math.cos(angle_y)
        sin_y = math.sin(angle_y)

        new_x = x * cos_y - z * sin_y
        new_z = x * sin_y + z * cos_y

        final_x = new_x
        final_y = y * cos_x - new_z * sin_x
        final_z = y * sin_x + new_z * cos_x

        return final_x, final_y, final_z

    def project(self, point):
        x, y, z = point
        scale = 200 / (200 + z)  # Simple perspective projection
        screen_x = x * scale + 200
        screen_y = y * scale + 200
        return screen_x, screen_y

    def animate(self):
        start_time = time.time()

        self.canvas.delete("all")
        self.angle_x += 0.01
        self.angle_y += 0.01

        rotated_vertices = [self.rotate_point(vertex, self.angle_x, self.angle_y) for vertex in self.cube_vertices]
        projected_vertices = [self.project(vertex) for vertex in rotated_vertices]

        for edge in self.cube_edges:
            p1, p2 = edge
            x1, y1 = projected_vertices[p1]
            x2, y2 = projected_vertices[p2]
            self.canvas.create_line(x1, y1, x2, y2)

        end_time = time.time()
        frame_time = (end_time - start_time) * 1000  # Convert to milliseconds

        # Calculate FPS and display both FPS and average frame time
        self.frames += 1
        current_time = time.time()
        elapsed_time = current_time - self.last_frame_time
        if elapsed_time >= 1:  # Update FPS and average frame time every second
            fps = self.frames / elapsed_time
            avg_frame_time = elapsed_time / self.frames * 1000  # Average frame time in milliseconds
            print(f"\rFPS: {fps:.4f} - Avg Frame Time: {avg_frame_time:.6f} ms", end="")
            self.frames = 0
            self.last_frame_time = current_time

        self.root.after(10, self.animate)

if __name__ == "__main__":
    root = tk.Tk()
    app = SpinningCubeApp(root)
    root.mainloop()


ppm.py
import time

def main():
    while True:
        def help_command():
            file_path = 'source/help_ppm.txt'

            try:
                # Open the file in read mode
                with open(file_path, 'r') as file:
                    # Read the contents of the file
                    file_contents = file.read()
                    print(file_contents)

            except FileNotFoundError:
                print(f"The file '{file_path}' does not exist.")
            except Exception as e:
                print(f"An error occurred: {e}")

        def default_command():
            command = input(">> ")

            if command == ".help":
                help_command()
            elif command == ".exit":
                quit()
            else:
                print("error: '" + command + "' isn't a command")
        default_command()

if __name__ == "__main__":
    main()


main.py
import sys
import threading
import subprocess
import socket

import time
import string
import random

# ppm python

# welcome
print("Welcome to terminal clone v2.14")
print("type '.help' for more information!")


def main():
    # login
    global username
    global password
    global logged_in
    logged_in = False

    global internet_state
    internet_state = False

    while True:
        # login system
        def register_user(username, password):
            with open('source/users.txt', 'a') as file:
                file.write(f"{username}:{password}\n")

        def check_credentials(username, password):
            with open('source/users.txt', 'r') as file:
                for line in file:
                    stored_username, stored_password = line.strip().split(':')
                    if stored_username == username and stored_password == password:
                        return True
            return False

        # text color
        class Colors:
            RESET = "\033[0m"
            BLACK = "\033[30m"
            RED = "\033[31m"
            GREEN = "\033[32m"
            YELLOW = "\033[33m"
            BLUE = "\033[34m"
            MAGENTA = "\033[35m"
            CYAN = "\033[36m"
            WHITE = "\033[37m"
            BOLD = "\033[1m"
            UNDERLINE = "\033[4m"

        def color_text(text, color_code):
            return f"{color_code}{text}{Colors.RESET}"

        # Loading & Animation
        def spinning_text(text):
            spinner = ['|', '/', '-', '\\']
            spin_round = random.randint(10, 40)
            for i in range(spin_round):
                for char in spinner:
                    print(f"\r{char} {text}", end="")
                    time.sleep(0.1)

        def loading_bar(iterations, bar_length=50, text="Loading"):
            for i in range(iterations + 1):
                delay = random.randint(1, 6)
                percent = i / iterations
                hashes = '#' * int(round(percent * bar_length))
                spaces = '-' * (bar_length - len(hashes))
                percent_str = f"{int(percent * 100)}%"
                custom_text = f"{text} [{hashes}{spaces}] {percent_str}"
                print(f"\r{custom_text}", end='', flush=True)
                time.sleep(delay / 10)

        def countdown_timer(seconds, custom_text):
            for remaining in range(seconds, -1, -1):
                print(f"{custom_text} {remaining} seconds", end='\r')
                time.sleep(1)

        # Command process

        # system test
        process_name = ["System", "dwm.exe", "smss.exe", "csrss.exe", "lsass.exe", "svchost.exe", "conhost.exe",
                        "spoolsv.exe", "wininit.exe", "taskhost.exe", "explorer.exe", "services.exe", "winlogon.exe", "System Idle Process"]

        def process_list(text, process_name):
            for text_list in process_name:
                print(f"\r{text} {text_list}", end="")
                delay = random.randint(1, 10)
                time.sleep(delay/10)

        def encryption(length):
            characters = string.ascii_uppercase + string.ascii_lowercase + string.digits
            random_characters = ''.join(random.choice(
                characters) for _ in range(length))
            return random_characters

        # all process display
        def system_test():
            # start timer
            start_time = time.time()

            print(color_text("DO NOT CLOSE THIS WINDOWS!!!", Colors.RED))
            time.sleep(2.5)
            print(color_text("This process will take a few minute", Colors.YELLOW))
            time.sleep(5)
            spinning_text("Preparing for test...")

            # file check
            loading_bar(100, bar_length=50, text="Checking system file")
            print("")

            # encryption test
            spinning_text("Preparing for encryption test...")
            for i in range(1000):
                random_combination = encryption(20)
                print(f"\rEncryption Key:", random_combination)
                delay = random.randint(1, 10)
                time.sleep(delay/100)

            # process test
            process_list("Checking system process", process_name)

            # timer stop
            end_time = time.time()
            elapsed_time = end_time - start_time

            # finish command
            print(
                f"\r{color_text('Checking system Complete', Colors.GREEN)} - Elapsed Time: {elapsed_time:.6f} seconds")

        # internet check
        def internet():
            global internet_state

            def internet_connect():
                try:
                    return True
                except OSError:
                    pass
                return False

            if internet_connect():
                time.sleep(2)
                spinning_text("connecting to server...")
                print(color_text("\nConnecting complete", Colors.GREEN))
                internet_state = True
            else:
                spinning_text("connecting...")
                print("\nconnecting fail...")
                countdown_timer(5, "cannot connect to server, try again in:")
                time.sleep(2)
                print("")
                connect = input("Type 'Y' to reconnect or 'n' to cancel: ")
                if connect == "Y":
                    internet()
                elif connect == "n":
                    pass

        # render 3d
        def render():
            script_path = '3d.py'
            return_code = subprocess.call(['python', script_path])

            if return_code != 0:
                print(
                    f"Error while running the script: Command '['python', '{script_path}']' returned non-zero exit status {return_code}")
            print("")

        # ppm install
        def ppm():
            script_path = 'ppm.py'
            subprocess.Popen(
                ['start', 'cmd', '/k', 'python', script_path], shell=True)

        # help
        def help_command():
            file_path = 'source/help.txt'

            try:
                # Open the file in read mode
                with open(file_path, 'r') as file:
                    # Read the contents of the file
                    file_contents = file.read()
                    print(file_contents)

            except FileNotFoundError:
                print(f"The file '{file_path}' does not exist.")
            except Exception as e:
                print(f"An error occurred: {e}")

        def about():
            file_path = 'source/about.txt'

            try:
                # Open the file in read mode
                with open(file_path, 'r') as file:
                    # Read the contents of the file
                    file_contents = file.read()
                    print(file_contents)

            except FileNotFoundError:
                print(f"The file '{file_path}' does not exist.")
            except Exception as e:
                print(f"An error occurred: {e}")

        # contect
        def contect():
            link_fb = "https://www.facebook.com/ShadowWSsnow/"
            link_ig = "https://www.instagram.com/shadoww_ssnow/"
            link_rd = "https://www.reddit.com/user/ShadowW_Ssnow/"
            link_gh = "https://github.com/donut9164/"

            fb_text = "Facebook   : "
            ig_text = "Instragram : "
            rd_text = "Reddit     : "
            gh_text = "github     : "

            fb = f"{fb_text}{link_fb}"
            ig = f"{ig_text}{link_ig}"
            rd = f"{rd_text}{link_rd}"
            gh = f"{gh_text}{link_gh}"

            print(fb)
            print(ig)
            print(rd)
            print(gh)

        # login
        def register():
            global logged_in

            if not internet_state:
                print("Please connect to server before register")
            else:
                username = input(f"\rEnter username: ")
                password = input(f"\rEnter password: ")
                register_user(username, password)
                loading_bar(2, bar_length=50, text="Creating new profile")
                print("\nUser registered successfully!\n")

        def login():
            global logged_in

            if not internet_state:
                print("Please connect to server before login")
            else:
                username = input(f"\rEnter username: ")
                password = input(f"\rEnter password: ")

                if check_credentials(username, password):
                    logged_in = True
                    print("Login as " + "[" + username + "]")
                else:
                    print("Invalid username or password!\n")

        # main

        def default_command():
            global username
            global password
            global internet_state
            global logged_in
            username = ""

            command = input(">> ")
            
            if command == "/register":
                if not logged_in:
                    register()
                else:
                    print("Please logout before creating a new account")

            elif command == "/login":
                if not logged_in:
                    login()
                else:
                    print("Please logout before login a new account")

            elif command == "/logout":
                if logged_in:
                    logged_in = False
                    print("Logged out successfully!\n")
                else:
                    print("You didn't login type:")
                    print("/register to create new account")
                    print("/login    to login into your account")

            elif command == "/state":
                if not logged_in:
                    print("You didn't login type:")
                    print("/register to create new account")
                    print("/login    to login into your account")
                elif logged_in:
                    print("You login as " + "[" + username + "]")

            # internet command
            elif command == ".connect":
                internet()
            # system command
            elif command == ".help":
                help_command()
            elif command == ".test":
                system_test()
            elif command == ".render":
                render()
            elif command == "ppm":
                if internet_state:
                    if logged_in:
                        ppm()
                    else:
                        print("Please login before open ppm install package")

                else:
                    print(
                        "Please connect to server before open ppm install package")
                print("")

            elif command == ".exit":
                quit()

            # about
            elif command == ".about":
                about()
            elif command == ".contect":
                contect()
            else:
                print("error: '" + command + "' isn't a command")
        default_command()


if __name__ == "__main__":
    main()
